File: kash_kash_app/lib/domain/entities/quest.dart
/// Quest difficulty levels
enum QuestDifficulty { easy, medium, hard, expert }

/// Location types for quests
enum LocationType { city, forest, park, water, mountain, indoor }

/// Quest entity representing a geocaching location
class Quest {
  final String id;
  final String title;
  final String? description;
  final double latitude;
  final double longitude;
  final double radiusMeters;
  final String createdBy;
  final bool published;
  final QuestDifficulty? difficulty;
  final LocationType? locationType;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? syncedAt;

  const Quest({
    required this.id,
    required this.title,
    this.description,
    required this.latitude,
    required this.longitude,
    this.radiusMeters = 3.0,
    required this.createdBy,
    this.published = false,
    this.difficulty,
    this.locationType,
    required this.createdAt,
    required this.updatedAt,
    this.syncedAt,
  });

  Quest copyWith({
    String? id,
    String? title,
    String? description,
    double? latitude,
    double? longitude,
    double? radiusMeters,
    String? createdBy,
    bool? published,
    QuestDifficulty? difficulty,
    LocationType? locationType,
    DateTime? createdAt,
    DateTime? updatedAt,
    DateTime? syncedAt,
  }) {
    return Quest(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      radiusMeters: radiusMeters ?? this.radiusMeters,
      createdBy: createdBy ?? this.createdBy,
      published: published ?? this.published,
      difficulty: difficulty ?? this.difficulty,
      locationType: locationType ?? this.locationType,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      syncedAt: syncedAt ?? this.syncedAt,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Quest && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;
}

File: backend/src/Entity/Quest.php
<?php

namespace App\Entity;

use ApiPlatform\Metadata\ApiResource;
use ApiPlatform\Metadata\Delete;
use ApiPlatform\Metadata\Get;
use ApiPlatform\Metadata\GetCollection;
use ApiPlatform\Metadata\Post;
use ApiPlatform\Metadata\Put;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Uid\Uuid;
use Symfony\Component\Validator\Constraints as Assert;

#[ORM\Entity]
#[ORM\Table(name: 'quests')]
#[ApiResource(
    operations: [
        new Get(),
        new GetCollection(),
        new Post(security: "is_granted('ROLE_ADMIN')"),
        new Put(security: "is_granted('ROLE_ADMIN')"),
        new Delete(security: "is_granted('ROLE_ADMIN')"),
    ]
)]
class Quest
{
    public const DIFFICULTY_EASY = 'easy';
    public const DIFFICULTY_MEDIUM = 'medium';
    public const DIFFICULTY_HARD = 'hard';
    public const DIFFICULTY_EXPERT = 'expert';

    public const LOCATION_CITY = 'city';
    public const LOCATION_FOREST = 'forest';
    public const LOCATION_PARK = 'park';
    public const LOCATION_WATER = 'water';
    public const LOCATION_MOUNTAIN = 'mountain';
    public const LOCATION_INDOOR = 'indoor';

    #[ORM\Id]
    #[ORM\Column(type: 'uuid', unique: true)]
    private Uuid $id;

    #[ORM\Column(length: 255)]
    #[Assert\NotBlank]
    private string $title;

    #[ORM\Column(type: Types::TEXT, nullable: true)]
    private ?string $description = null;

    #[ORM\Column(type: Types::FLOAT)]
    #[Assert\Range(min: -90, max: 90)]
    private float $latitude;

    #[ORM\Column(type: Types::FLOAT)]
    #[Assert\Range(min: -180, max: 180)]
    private float $longitude;

    #[ORM\Column(type: Types::FLOAT)]
    private float $radiusMeters = 3.0;

    #[ORM\ManyToOne(targetEntity: User::class, inversedBy: 'createdQuests')]
    #[ORM\JoinColumn(nullable: false)]
    private User $createdBy;

    #[ORM\Column(type: Types::BOOLEAN)]
    private bool $published = false;

    #[ORM\Column(length: 20, nullable: true)]
    private ?string $difficulty = null;

    #[ORM\Column(length: 20, nullable: true)]
    private ?string $locationType = null;

    #[ORM\Column(type: Types::DATETIME_IMMUTABLE)]
    private \DateTimeImmutable $createdAt;

    #[ORM\Column(type: Types::DATETIME_IMMUTABLE)]
    private \DateTimeImmutable $updatedAt;

    #[ORM\OneToMany(targetEntity: QuestAttempt::class, mappedBy: 'quest')]
    private Collection $attempts;

    public function __construct()
    {
        $this->id = Uuid::v4();
        $this->createdAt = new \DateTimeImmutable();
        $this->updatedAt = new \DateTimeImmutable();
        $this->attempts = new ArrayCollection();
    }

    public function getId(): Uuid
    {
        return $this->id;
    }

    public function getTitle(): string
    {
        return $this->title;
    }

    public function setTitle(string $title): static
    {
        $this->title = $title;
        return $this;
    }

    public function getDescription(): ?string
    {
        return $this->description;
    }

    public function setDescription(?string $description): static
    {
        $this->description = $description;
        return $this;
    }

    public function getLatitude(): float
    {
        return $this->latitude;
    }

    public function setLatitude(float $latitude): static
    {
        $this->latitude = $latitude;
        return $this;
    }

    public function getLongitude(): float
    {
        return $this->longitude;
    }

    public function setLongitude(float $longitude): static
    {
        $this->longitude = $longitude;
        return $this;
    }

    public function getRadiusMeters(): float
    {
        return $this->radiusMeters;
    }

    public function setRadiusMeters(float $radiusMeters): static
    {
        $this->radiusMeters = $radiusMeters;
        return $this;
    }

    public function getCreatedBy(): User
    {
        return $this->createdBy;
    }

    public function setCreatedBy(User $createdBy): static
    {
        $this->createdBy = $createdBy;
        return $this;
    }

    public function isPublished(): bool
    {
        return $this->published;
    }

    public function setPublished(bool $published): static
    {
        $this->published = $published;
        return $this;
    }

    public function getDifficulty(): ?string
    {
        return $this->difficulty;
    }

    public function setDifficulty(?string $difficulty): static
    {
        $this->difficulty = $difficulty;
        return $this;
    }

    public function getLocationType(): ?string
    {
        return $this->locationType;
    }

    public function setLocationType(?string $locationType): static
    {
        $this->locationType = $locationType;
        return $this;
    }

    public function getCreatedAt(): \DateTimeImmutable
    {
        return $this->createdAt;
    }

    public function getUpdatedAt(): \DateTimeImmutable
    {
        return $this->updatedAt;
    }

    public function setUpdatedAt(\DateTimeImmutable $updatedAt): static
    {
        $this->updatedAt = $updatedAt;
        return $this;
    }

    public function getAttempts(): Collection
    {
        return $this->attempts;
    }
}
